### PR TITLE
Add colour Labels to elements 

### LINKED ISSUE #1785: color labels
## Affected Area
Eidtor Interface

## Requested Feature
Use colors to label individual elements (Articles, Blocks, Components). Similar to Finder in MacOSX.
Color might be selected by the user, or from a list of predefined colors.

## Use Case
Color labels would be useful to highlight sections when editing long pages.

## Current Workaround
None

### PR DESCRIPTION
**Pull request by lc-thomasberger**
fixes #1785

### CODE REVIEW

**Comment by canstudios-louisem** [comment]
If I set a colour label on an article or block and then refresh the label stays if I do the same on a component it does not.

**Comment by lc-thomasberger** [comment]
@canstudios-louisem `_colorLabel` was not defined in the schema for components ... thought it would also inherit from content. Should work now.

**Comment by taylortom** [comment]
Works for well me :) 

A few minor things:
- Could you switch the context label to use the british english variation 'colour' (not sure if we have this documented in a language style guide anywhere, but that's what we use)
- I think it would also be useful to add some kind of description text/title in the colour select popup, as it looks a bit blank/not obvious what's going on (see screen below)
- One last thing: should have some way to configure these colours other than in the LESS? Seems like something people might want to change

screen shot 2018-03-23 at 15 31 56

**Comment by lc-thomasberger** [comment]
Cool!
The re-design of the labels required switching to css rather than inline styling. So the colors must be set in the theme of the AT. The easiest way would be a file that stores all the variables that can than be customized. I will updated the colors to use generic names instead of the colors.

**Comment by canstudios-louisem** [comment]
If you're going to put them in a file can they just go in the config? I would rather we didn't create more config files.

**Comment by lc-thomasberger** [comment]
@canstudios-louisem it would be a less file. If it should be set via configuration we would need to inject the file in the grunt task.

**Comment by dancgray** [comment]
Colour labels work really well. Is anyone else getting an error "constructorFn is not a constructor" when trying to edit a component? Could be my install.

**Comment by lc-thomasberger** [comment]
@dancgray good catch ... I will have a look at this. It only shows a popup without an error so not sure what causes this

File: frontend/src/core/less/colourLabels.less [hunk 1]
```diff
@@ -0,0 +1,14 @@
+@colorlabel-1: #616161;
```
Comment (tomgreenfield) [inline frontend/src/core/less/colourLabels.less]: Name of this file is inconsistent with the other Less files.

File: frontend/src/modules/colorLabel/views/colorLabelPopupView.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            if (!this.selected) return;
+            this.addItem();
+        },
+
+        preRender: function() {
+            this.selected = -1;
+        },
+
+        postRender: function() {
+            if (!this.parentView.model.has('_colorLabel')) return;
```
Comment (tomgreenfield) [inline frontend/src/modules/colorLabel/views/colorLabelPopupView.js]: No need for `.has()`.
```js
var color = this.parentView.model.get('_colorLabel');

if (!color) return;

this.selected = color;
this.updateClasses();
```

File: frontend/src/modules/colorLabel/views/colorLabelPopupView.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            });
+        },
+
+        onApply: function(event) {
+            if (!this.selected) return;
+            this.addItem();
+        },
+
+        preRender: function() {
+            this.selected = -1;
```
Comment (tomgreenfield) [inline frontend/src/modules/colorLabel/views/colorLabelPopupView.js]: `this.selected` could be `undefined` (on initialise), [`null`](https://github.com/adaptlearning/adapt_authoring/pull/1921/files#diff-4082fa8f4c93488e4a000c7ea57ed2dfR35) or [`-1`](https://github.com/adaptlearning/adapt_authoring/pull/1921/files#diff-4082fa8f4c93488e4a000c7ea57ed2dfR50), when there's no colour chosen.

Is there a reason for the three possible values?

Might be easier to have `selected: null` on the view and set / return to `null` when appropriate.

File: frontend/src/modules/editor/global/views/editorOriginView.js [hunk 4]
```diff
@@ -10,6 +10,14 @@ define(function(require){
       'click .field-object .legend': 'onFieldObjectClicked'
     },
 
+    attributes: function() {
+      if (this.model && this.model.has('_colorLabel') && this.model.get('_colorLabel')) {
```
Comment (tomgreenfield) [inline frontend/src/modules/editor/global/views/editorOriginView.js]: No need for `.has()`.
```js
var colorLabel = this.model && this.model.get('_colorLabel');

if (colorLabel) return { 'data-colorlabel': colorLabel };
```

File: frontend/src/modules/colorLabel/views/colorLabelPopupView.js (hunk 3 above)
Comment (lc-thomasberger) [reply frontend/src/modules/colorLabel/views/colorLabelPopupView.js]: I thought about allowing custom colors first ... will have to revisit.

**Comment by lc-thomasberger** [comment]
Thanks @tomgreenfield added your suggestions as they improve the code.

File: frontend/src/modules/colorLabel/views/colorLabelPopupView.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            if (!color) return;
+            this.selected = color;
+            this.updateClasses();
+        },
+
+        applyOnParent: function() {
+            var color = this.selected;
+
+            if (this.selected) {
+                this.parentView.$el.attr('data-colorlabel', color);
```
Comment (tomgreenfield) [inline frontend/src/modules/colorLabel/views/colorLabelPopupView.js]: Now you chosen `null` as the falsey value, this whole function be condensed into one line:
```js
this.parentView.$el.attr('data-colorlabel', this.selected);
```
Reference: http://api.jquery.com/attr/#attr-attributeName-value.

File: frontend/src/modules/colorLabel/views/colorLabelPopupView.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            });
+        },
+
+        onApply: function(event) {
+            if (!this.selected) return;
+            this.addItem();
+        },
+
+        preRender: function() {
+            this.selected = null;
```
Comment (tomgreenfield) [inline frontend/src/modules/colorLabel/views/colorLabelPopupView.js]: `this.selected` can be declared on instantiation if you preferred to not use this function for it, like [`className`](https://github.com/adaptlearning/adapt_authoring/pull/1921/files#diff-4082fa8f4c93488e4a000c7ea57ed2dfR8) is:
```js
selected: null
```

**Comment by tomgreenfield** [comment]
@lc-thomasberger, have you pushed your recent commit(s)?

File: routes/lang/en-application.json [hunk 7]
```diff
@@ -311,5 +313,8 @@
   "app.article": "article",
   "app.block": "block",
   "app.component": "component",
-  "app.errorloadconfig": "Failed to load configuration settings for %{course}"
+  "app.errorloadconfig": "Failed to load configuration settings for %{course}",
+  "app.colourLabel": "Colour Label",
+  "app.colourLabelTitle": "Colour Label",
+  "app.colourLabelIntro": "The colours are <b>only</b> applied in the authoring tool and will <b>not affect the genereated course</b>. The settings are synced with all users in the authoring tool."
```
Comment (tomgreenfield) [inline routes/lang/en-application.json]: Should be `generated`.

**Pull request merged**